### DISCUSSION TITLE
Schema

### DISCUSSION THREAD

**Discussion by afflom** [emporous/community#6] [Design]
## Control Signalling
UOR Schema is a special Collection type that provides the control mechanisms of the UOR Framework. UOR Schema Collections have the following control mechanisms. 

1. Attribute type declarations
2. Entrypoint
3. Required attributes
4. Default attribute mapping
5. Event Engine declaration
6.  Event Engine command mappings
7. Default Execution Parameters

### 1. Attribute Type Declarations

UOR Attribute value types are declared within the Schema. UOR performs validations during Collection building. One of these validations is attribute type checking. This is performed to promote greater reliability of Event Engine processing by reducing occurrences of Event Engines receiving unexpected input. 

### 2. Entrypoint

An Entrypoint is an attribute address of the first nodes presented to an Event Engine when an Event Engine begins its processing. This is the first content passed to the Event Engine. 

### 3. Required Attributes

Used during Collection build, this is a list of attributes that are required to be present in a Collection. The attributes referenced by the Entrypoint are an extension of this list. 

### 4. Default Attribute Mapping

Default attribute mappings are a dataset config that is implicitly applied to a Collection's contents during building. The user defined dataset config is added to the default attribute mappings from the Schema. This is automatically referenced during Collection build to simplify the process of defining and building a dataset. 

### 5. Event Engine Declaration

An Event Engine Declaration is used to import an Event Engine into a Schema and referenced by the UOR Client during Event Lookup. 

### 6. Event Engine Command Mappings

These mappings are used to enumerate the commands exposed by attribute control signalling. These could be referenced by the user in a help subcommand workflow or used to dynamically configure the Event Engine proxy (proposed in #3). 

### 7. Default Execution Parameters

Default Execution Parameters are the (override-able) isolation/runtime parameters needed by the imported Event Engine during normal operation. These parameters convey capabilities given to the Event Engine Execution environment such as network, device, and other host capabilities. The user should be prompted if an Event Engine requires special capabilities. 

##
Please consider reviewing #4 and #5 for more information about Collections, Schema, and Event Engines. Also, please let me know if I've overlooked anything or if you have any questions.

**Comment by jpower432** [comment]
Hey @afflom

One thing about this proposal that I think we need to be mindful of is key collision between the default attribute mapping and the user mapping. The idea of required attributes and the default attribute mapping seems to overlap. One would assume that the necessary attributes and the default mapping have the required keys. 

This leaves me with two questions:

1. Would it make more sense to have the required values along with the required keys in one section? 
2. Is the point of the default attribute mapping to provide default values when the user does not specify the required keys?

**Reply by afflom** [reply]
I had originally referred to the list in my post as Schema sections, but backtracked and referred to them as mechanisms. That is an out for me to refer to these elements outside of where they will ultimately live, because I was concerned about getting implementation details wrong. 

Default mappings are to ease the burden of the content publisher when interacting with attributes, especially when defining in-content links. Theoretically, a user could use normal relative paths to link to adjacent content (within a collection) as they normally would, and the additional attributes on objects would implicitly form deeper references to underlying attributes that are defined in the schema.
